services/pagerduty2/service.go

```diff
@@ -0,0 +1,388 @@
+	c := s.config()
+	if !c.Enabled {
+		return "", nil, errors.New("service is not enabled")
+	}
+
+	var severity string
+	eventType := "trigger"
+
+	switch level {
+	case alert.Warning:
+		severity = "warning"
+	case alert.Critical:
+		severity = "critical"
+	case alert.Info:
+		severity = "info"
+	default:
+		// default is a 'resolve' function
+		return s.sendResolve(c, serviceKey, incidentKey)
+	}
+
+	// create a new AlertPayload for us to fire off
+	ap := &AlertPayload{
+		Payload: &PDCEF{},
+	}
+
+	if serviceKey == "" {
+		ap.RoutingKey = c.ServiceKey
+	} else {
+		ap.RoutingKey = serviceKey
+	}
+
+	ap.Client = "kapacitor"
+	ap.ClientURL = s.HTTPDService.URL()
+	ap.DedupKey = incidentKey
+	ap.EventAction = eventType
+
+	ap.Payload.CustomDetails = make(map[string]interface{})
+	if err := json.Unmarshal([]byte(html.UnescapeString(details)), &ap.Payload.CustomDetails); err != nil {
+		// Likely an invalid json string, so let's JSONify it
+		str := html.UnescapeString(fmt.Sprintf(`{"raw_details": "%s"}`, details))
+
+		if err := json.Unmarshal([]byte(str), &ap.Payload.CustomDetails); err != nil {
+			return "", nil, err
+		}
+	}
+
+	// The API doesn't explicitly mention a requirement for nanosecond resolution but payloads seem to
+	// fail if we don't include it (even zeroes). This hack is not graceful, but adds a negligible
+	// nanosecond resolution to our timestamp
+	m := timestamp.Format(time.RFC3339Nano)
+	if timestamp.Nanosecond() == 0 {
+		m = time.Unix(timestamp.Unix(), 1).In(timestamp.Location()).Format(time.RFC3339Nano)
+	}
+
+	ap.Payload.Class = data.TaskName
+	ap.Payload.Severity = severity
+	ap.Payload.Source = "unknown"
+	ap.Payload.Summary = desc
+	ap.Payload.Timestamp = m
+
+	if _, ok := data.Tags["host"]; ok {
+		ap.Payload.Source = data.Tags["host"]
+	}
+
+	// Post data to PagerDuty
+	var post bytes.Buffer
+	enc := json.NewEncoder(&post)
+	err := enc.Encode(ap)
+	if err != nil {
+		return "", nil, err
+	}
+
+	return c.URL, &post, nil
+}
+
+// HandlerConfig defines the high-level struct required to connect to PagerDuty
+type HandlerConfig struct {
+	// The service key to use for the alert.
+	// Defaults to the value in the configuration if empty.
+	ServiceKey string `mapstructure:"service-key"`
+}
+
+type handler struct {
+	s    *Service
+	c    HandlerConfig
+	diag Diagnostic
+}
+
+// Handler is a bound method to the Service struct that returns the appropriate alert handler for PagerDuty
+func (s *Service) Handler(c HandlerConfig, ctx ...keyvalue.T) alert.Handler {
+	return &handler{
+		s:    s,
+		c:    c,
+		diag: s.diag.WithContext(ctx...),
+	}
+}
+
+// Handle is a bound method to the handler that processes a given alert
+func (h *handler) Handle(event alert.Event) {
+	if err := h.s.Alert(
+		h.c.ServiceKey,
+		event.State.ID,
+		event.State.Message,
+		event.State.Level,
+		event.State.Details,
```

Comment:
As per the above comment we can probably just remove passing the details down if they are not going to be used.